pallets/dca/src/lib.rs

```diff
@@ -0,0 +1,995 @@
+			Error::<T>::TotalAmountShouldBeLargerThanStorageBond
+		);
+
+		Ok(())
+	}
+
+	fn ensure_schedule_is_suspended(schedule_id: ScheduleId) -> DispatchResult {
+		ensure!(
+			Suspended::<T>::contains_key(schedule_id),
+			Error::<T>::ScheduleMustBeSuspended
+		);
+
+		Ok(())
+	}
+
+	fn ensure_schedule_exists(schedule_id: &ScheduleId) -> DispatchResult {
+		ensure!(Schedules::<T>::contains_key(schedule_id), Error::<T>::ScheduleNotExist);
+
+		Ok(())
+	}
+
+	fn ensure_origin_is_schedule_owner(schedule_id: ScheduleId, who: &T::AccountId) -> DispatchResult {
+		let schedule = Schedules::<T>::get(schedule_id).ok_or(Error::<T>::ScheduleNotExist)?;
+		ensure!(*who == schedule.owner, Error::<T>::Forbidden);
+
+		Ok(())
+	}
+
+	fn ensure_sell_amount_is_bigger_than_transaction_fee(
+		schedule: &Schedule<T::AccountId, T::Asset, T::BlockNumber>,
+	) -> DispatchResult {
+		match schedule.order {
+			Order::Sell {
+				asset_in, amount_in, ..
+			} => {
+				let transaction_fee = Self::get_transaction_fee(asset_in)?;
+				ensure!(amount_in > transaction_fee, Error::<T>::TradeAmountIsLessThanFee);
+			}
+			Order::Buy { .. } => {
+				//For buy we don't check as the calculated amount in will always include the fee
+			}
+		}
+
+		Ok(())
+	}
+
+	pub fn execute_schedule(current_blocknumber: T::BlockNumber, weight: &mut u64, schedule_id: ScheduleId) {
+		*weight += Self::get_execute_schedule_weight();
+
+		let schedule = exec_or_return_if_none!(Schedules::<T>::get(schedule_id));
+		let origin: OriginFor<T> = Origin::<T>::Signed(schedule.owner.clone()).into();
+		let blocknumber_for_schedule = exec_or_return_if_none!(current_blocknumber.checked_add(&schedule.period));
+
+		let sold_currency = schedule.order.get_asset_in();
+		if exec_or_return_if_err!(Self::price_change_is_bigger_than_max_allowed(
+			sold_currency,
+			schedule.order.get_asset_out()
+		)) {
+			exec_or_return_if_err!(Self::plan_schedule_for_block(blocknumber_for_schedule, schedule_id));
+			return;
+		}
+
+		let amount_to_unreserve = exec_or_return_if_err!(Self::amount_to_unreserve(&schedule.order));
+		let remaining_amount_to_use = exec_or_return_if_none!(RemainingAmounts::<T>::get(schedule_id));
+
+		T::Currency::unreserve_named(
+			&NAMED_RESERVE_ID,
+			sold_currency.into(),
+			&schedule.owner,
+			amount_to_unreserve.into(),
+		);
+
+		exec_or_return_if_err!(Self::decrease_remaining_amount(schedule_id, amount_to_unreserve));
+
+		if remaining_amount_to_use < amount_to_unreserve {
+			Self::complete_dca(&schedule.owner, schedule_id);
+			return;
+		}
+
+		exec_or_return_if_err!(Self::take_transaction_fee_from_user(&schedule.owner, &schedule.order));
+		let trade_result = Self::execute_trade(origin, &schedule.order);
+
+		match trade_result {
+			Ok(_) => {
+				exec_or_return_if_err!(Self::plan_schedule_for_block(blocknumber_for_schedule, schedule_id));
+			}
+			_ => {
+				exec_or_return_if_err!(Self::suspend_schedule(&schedule.owner, schedule_id));
+			}
+		}
```

Comment:
again here, regarding the macros. 

just by looking at this match statement - i cannot really tell what is going to happen on first sight. It should be possible to make it more clear.